src/geometry/manifold/ManifoldGeometry.cc

```diff
@@ -106,11 +113,80 @@ std::shared_ptr<const PolySet> ManifoldGeometry::toPolySet() const {
         mesh.vertProperties[i],
         mesh.vertProperties[i + 1],
         mesh.vertProperties[i + 2]);
-  for (size_t i = 0; i < mesh.triVerts.size(); i += 3)
-    ps->indices.push_back({
-        static_cast<int>(mesh.triVerts[i]),
-        static_cast<int>(mesh.triVerts[i + 1]),
-        static_cast<int>(mesh.triVerts[i + 2])});
+
+  if (Feature::ExperimentalRenderColors.is_enabled()) {
+    ps->colors.reserve(originalIDToColor_.size());
+    ps->color_indices.reserve(ps->indices.size());
+
+    auto colorScheme = ColorMap::inst()->findColorScheme(RenderSettings::inst()->colorscheme);
+    int32_t faceFrontColorIndex = -1;
+    int32_t faceBackColorIndex = -1;
+
+    std::map<Color4f, int32_t> colorToIndex;
+    std::map<uint32_t, int32_t> originalIDToColorIndex;
+
+    auto getFaceFrontColorIndex = [&]() -> int {
+      if (faceFrontColorIndex < 0) {
+        faceFrontColorIndex = ps->colors.size();
+        ps->colors.push_back(ColorMap::getColor(*colorScheme, RenderColor::CGAL_FACE_FRONT_COLOR));
+      }
+      return faceFrontColorIndex;
+    };
+    auto getFaceBackColorIndex = [&]() -> int {
+      if (faceBackColorIndex < 0) {
+        faceBackColorIndex = ps->colors.size();
+        ps->colors.push_back(ColorMap::getColor(*colorScheme, RenderColor::CGAL_FACE_BACK_COLOR));
+      }
+      return faceBackColorIndex;
+    };
+
+    auto getColorIndex = [&](uint32_t originalID) -> int32_t {
+      if (subtractedIDs_.find(originalID) != subtractedIDs_.end()) {
+        return getFaceBackColorIndex();
+      }
+      auto colorIndexIt = originalIDToColorIndex.find(originalID);
+      if (colorIndexIt != originalIDToColorIndex.end()) {
+        return colorIndexIt->second;
+      }
+      auto colorIt = originalIDToColor_.find(originalID);
+      if (colorIt == originalIDToColor_.end()) {
+        return getFaceFrontColorIndex();
+      }
+      const auto & color = colorIt->second;
+      
+      auto pair = colorToIndex.insert({color, ps->colors.size()});
+      if (pair.second) {
+        ps->colors.push_back(color);
+      }
+      int32_t color_index = pair.first->second;
+      originalIDToColorIndex[originalID] = color_index;
+      return color_index;
+    };
+
+    auto start = mesh.runIndex[0];
+    for (int run = 0, numRun = mesh.runIndex.size() - 1; run < numRun; ++run) {
+      const auto id = mesh.runOriginalID[run];
+      const auto end = mesh.runIndex[run + 1];
+      const size_t numTri = (end - start) / 3;
+      assert(numTri > 0);
```

Comment:
Reading Manifold source code, It looks like it's perfectly valid that `runIndex` entries have zero triangles, in cases where an object didn't contribute to the final mesh.

Reply:
I do wonder, though, why you're not seeing this assertion on your end?

Reply:
Oh right, I think I use that to ensure I get the combined transform for `show()` and `only()` akin to `#` and `*`.

Reply:
Couldn't reproduce this on Mac yet (w/ CGAL 5.4.5; I suppose different versions could matter here since it's about minkowski), my code in createManifoldFromTriangularPolySet might be dodgy.

Reply:
My test was on macOS w/CGAL 5.5 (built from source using scripts/macosx-build-dependencies.sh).

Reply:
Can also reproduce without using minkowski

Reply:
Just as a test, I removed the assert, and the above example produces some interesting artifacts:

<img width="809" alt="Screenshot 2024-07-01 at 21 11 46" src="https://github.com/openscad/openscad/assets/186094/5a2a9c24-40ff-468d-9c8e-2b213f6f742e">

Reply:
Is this on a (k)Intel Mac? Did you make sure the manifold submodule is updated?

Reply:
M1 mac - I'll double-check by updating and building from scratch

Reply:
Build from scratch; still same issue (assert and triangle coloring artifact)

Reply:
This example also behaves oddly on the web playground

Reply:
Some digging I realize my example was bogus (ambigous source triangles), so pls. disregard the screenshot above. However, the z tearing I see in the web playground it worth looking into.

Reply:
I managed to reproduce the assert failure on a fresh VM running macOS 14.4 and the latest Xcode.
I think the main difference from your setup may be that I've build the dependencies using scripts/macosx-build-dependencies.sh on both machines. I wonder if there is some minor difference in what libs get pulled in. ..or a bug surfacing in a 3rd party lib.

Reply:
I can also reproduce the assertion failure when using homebrew dependencies.
Not sure what next to try, perhaps you're on an older compiler? I'm on `Apple clang version 15.0.0 (clang-1500.1.0.2.5)`

Reply:
The web playground z buffer tearing feels like lazy union was used to produce two partially overlapping objects. If that's the case, it's WAI.

Reply:
Ah of course. Lazy union can’t currently be disabled, I need to expose the options somewhere.

Reply:
Yeah, so the remaining puzzling issue here is why the assertion failure is not consistent across environments.